### PR TITLE
CppCheck FanCoilUnits

### PR DESCRIPTION
**Pull request by rraustad**
Pull request overview
---------------------
 - Fixes CppCheck style suggestions
 - DESCRIBE PURPOSE OF THIS PULL REQUEST

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [ ] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [ ] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [ ] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

### CODE REVIEW

File: src/EnergyPlus/FanCoilUnits.cc [hunk 1]
```diff
@@ -308,22 +263,20 @@ namespace FanCoilUnits {
...
-            FanCoilNum = FanCoilIndex;
-
-            state.dataFanCoilUnits->FanCoilNumericFields(FanCoilNum).FieldNames.allocate(NumNumbers);
-            state.dataFanCoilUnits->FanCoilNumericFields(FanCoilNum).FieldNames = "";
-            state.dataFanCoilUnits->FanCoilNumericFields(FanCoilNum).FieldNames = cNumericFields;
+            state.dataFanCoilUnits->FanCoilNumericFields(FanCoilIndex).FieldNames.allocate(NumNumbers);
+            state.dataFanCoilUnits->FanCoilNumericFields(FanCoilIndex).FieldNames = "";
+            state.dataFanCoilUnits->FanCoilNumericFields(FanCoilIndex).FieldNames = cNumericFields;
 
             UtilityRoutines::IsNameEmpty(state, Alphas(1), CurrentModuleObject, ErrorsFound);
```
Comment (amirroth) [inline src/EnergyPlus/FanCoilUnits.cc]: Delete this?

File: src/EnergyPlus/FanCoilUnits.cc [hunk 2]
```diff
@@ -334,283 +287,266 @@ namespace FanCoilUnits {
                 UtilityRoutines::SameString(Alphas(3), "VariableFanVariableFlow") ||
                 UtilityRoutines::SameString(Alphas(3), "VariableFanConstantFlow") || UtilityRoutines::SameString(Alphas(3), "MultiSpeedFan") ||
                 UtilityRoutines::SameString(Alphas(3), "ASHRAE90VariableFan")) {
-                FanCoil(FanCoilNum).CapCtrlMeth = Alphas(3);
-                if (UtilityRoutines::SameString(Alphas(3), "ConstantFanVariableFlow")) FanCoil(FanCoilNum).CapCtrlMeth_Num = CCM::ConsFanVarFlow;
-                if (UtilityRoutines::SameString(Alphas(3), "CyclingFan")) FanCoil(FanCoilNum).CapCtrlMeth_Num = CCM::CycFan;
-                if (UtilityRoutines::SameString(Alphas(3), "VariableFanVariableFlow")) FanCoil(FanCoilNum).CapCtrlMeth_Num = CCM::VarFanVarFlow;
+                fanCoil.CapCtrlMeth = Alphas(3);
+                if (UtilityRoutines::SameString(Alphas(3), "ConstantFanVariableFlow")) fanCoil.CapCtrlMeth_Num = CCM::ConsFanVarFlow;
```
Comment (amirroth) [inline src/EnergyPlus/FanCoilUnits.cc]: enum with getEnumertionValue

File: src/EnergyPlus/FanCoilUnits.cc [hunk 3]
```diff
@@ -334,283 +287,266 @@ namespace FanCoilUnits {
...
+                        fanCoil.CoolCoilOutletNodeNum = WaterCoils::GetCoilOutletNode(state, fanCoil.CCoilType, fanCoil.CCoilName, IsNotOK);
                     } else {
-                        FanCoil(FanCoilNum).CoolCoilFluidInletNode = HVACHXAssistedCoolingCoil::GetCoilWaterInletNode(
-                            state, FanCoil(FanCoilNum).CCoilType, FanCoil(FanCoilNum).CCoilName, IsNotOK);
-                        FanCoil(FanCoilNum).CoolCoilInletNodeNum =
-                            HVACHXAssistedCoolingCoil::GetCoilInletNode(state, FanCoil(FanCoilNum).CCoilType, FanCoil(FanCoilNum).CCoilName, IsNotOK);
-                        FanCoil(FanCoilNum).CoolCoilOutletNodeNum = HVACHXAssistedCoolingCoil::GetCoilOutletNode(
-                            state, FanCoil(FanCoilNum).CCoilType, FanCoil(FanCoilNum).CCoilName, IsNotOK);
+                        fanCoil.CoolCoilFluidInletNode =
+                            HVACHXAssistedCoolingCoil::GetCoilWaterInletNode(state, fanCoil.CCoilType, fanCoil.CCoilName, IsNotOK);
```
Comment (amirroth) [inline src/EnergyPlus/FanCoilUnits.cc]: This pattern where you have to find the coil all over again in each one of these Get calls is pretty bad.

File: src/EnergyPlus/FanCoilUnits.hh [hunk 4]
```diff
@@ -109,158 +109,132 @@ namespace FanCoilUnits {
...
+        int HeatCoilOutletNodeNum = 0;         // index of heating coil outlet node number
+        int ControlZoneNum = 0;                // pointer to a zone served by a fancoil unit
+        int NodeNumOfControlledZone = 0;       // node number of controlled zone
+        bool ATMixerExists = false;            // True if there is an ATMixer
+        int ATMixerOutNode = 0;                // outlet air node number for the air terminal mixer
+        Real64 FanPartLoadRatio = 0.0;         // ratio of air flow to max air flow to simulation modulating fan
+        Real64 HeatCoilWaterFlowRatio = 0.0;   // ratio of water flow rate to max water flow rate
+        Real64 ControlZoneMassFlowFrac = 1.0;  // flow fraction of control zone (always 1 for zone equipment)
+        int MaxIterIndex = 0;                  // recurring message index
+        int RegulaFalsiFailedIndex = 0;        // iteration loop warning
```
Comment (Myoldmopar) [inline src/EnergyPlus/FanCoilUnits.hh:237]: I appreciate the eliminated constructor and better assignment initialization.

File: tst/EnergyPlus/unit/WaterCoils.unit.cc [hunk 5]
```diff
@@ -1599,7 +1599,7 @@ TEST_F(WaterCoilsTest, FanCoilCoolingWaterFlowTest)
     EXPECT_EQ(DataHVACGlobals::FanType_SimpleOnOff, state->dataFans->Fan(1).FanType_Num);
 
     GetFanCoilUnits(*state);
-    EXPECT_EQ("CONSTANTFANVARIABLEFLOW", state->dataFanCoilUnits->FanCoil(1).CapCtrlMeth);
+    EXPECT_TRUE(compare_enums(CCM::ConsFanVarFlow, state->dataFanCoilUnits->FanCoil(1).CapCtrlMeth_Num));
```
Comment (Myoldmopar) [inline tst/EnergyPlus/unit/WaterCoils.unit.cc:1602]: :+1:

File: src/EnergyPlus/FanCoilUnits.cc [hunk 6]
```diff
@@ -308,309 +263,291 @@ namespace FanCoilUnits {
...
                                                                      cNumericFields);
 
-            FanCoilNum = FanCoilIndex;
-
-            state.dataFanCoilUnits->FanCoilNumericFields(FanCoilNum).FieldNames.allocate(NumNumbers);
-            state.dataFanCoilUnits->FanCoilNumericFields(FanCoilNum).FieldNames = "";
-            state.dataFanCoilUnits->FanCoilNumericFields(FanCoilNum).FieldNames = cNumericFields;
+            state.dataFanCoilUnits->FanCoilNumericFields(FanCoilIndex).FieldNames.allocate(NumNumbers);
+            state.dataFanCoilUnits->FanCoilNumericFields(FanCoilIndex).FieldNames = "";
+            state.dataFanCoilUnits->FanCoilNumericFields(FanCoilIndex).FieldNames = cNumericFields;
```
Comment (Myoldmopar) [inline src/EnergyPlus/FanCoilUnits.cc:268]: I have a suspicion that we could eliminate this extra array of structs of arrays of field names pretty easily.  But OK for right now while addressing cppcheck things.

File: src/EnergyPlus/FanCoilUnits.cc [hunk 7]
```diff
@@ -2776,58 +2469,60 @@ namespace FanCoilUnits {
...
                 // get the maximum output of the fcu
                 Calc4PipeFanCoil(state, FanCoilNum, ControlledZoneNum, FirstHVACIteration, QUnitOutMax); // call without PLR means PLR = 1
 
                 if (QUnitOutMax < QZnReq) {
                     // more cooling than required, find reduced air and water flow rate to meet the load
                     // solve for the cold water flow rate with no limit set by flow rate lockdown
-                    auto f = [&state, FanCoilNum, FirstHVACIteration, ControlledZoneNum, &FanCoil, QZnReq](Real64 const PLR) {
-                        return CalcFanCoilPLRResidual(
-                            state, PLR, FanCoilNum, FirstHVACIteration, ControlledZoneNum, FanCoil(FanCoilNum).CoolCoilFluidInletNode, QZnReq);
+                    auto f = [&state, FanCoilNum, FirstHVACIteration, ControlledZoneNum, QZnReq](Real64 const PLR) {
```
Comment (Myoldmopar) [inline src/EnergyPlus/FanCoilUnits.cc:2517]: This change is just fine.  I guess it just revealed itself because of the change to using references.  :+1:

**Review by Myoldmopar — APPROVED** [review]
No issues here.  I am building it locally and found two build warnings that I cleaned up.  I'll push that up before merging assuming everything is happy.

**Comment by Myoldmopar** [comment]
CI was fully happy the last round aside from the noted build warnings.  I pushed a fix for those after local testing passed.  I'm merging this.  Thanks @rraustad .

**Pull request merged**